base_geo_route_optimization/models/geo_route_optimize.py

```diff
@@ -0,0 +1,189 @@
+        # Maximum number of rows that can be computed per request
+        max_rows = max_elements // num_addresses
+        # num_addresses = q * max_rows + r
+        q, r = divmod(num_addresses, max_rows)
+        dest_addresses = addresses
+        distance_matrix = []
+        # Send q requests, returning max_rows rows per request.
+        for i in range(q):
+            origin_addresses = addresses[i * max_rows: (i + 1) * max_rows]
+            response = self.send_request(origin_addresses, dest_addresses)
+            distance_matrix += self.build_distance_matrix(response)
+        # Get the remaining remaining r rows, if necessary.
+        if r > 0:
+            origin_addresses = addresses[q * max_rows: q * max_rows + r]
+            response = self.send_request(origin_addresses, dest_addresses)
+            distance_matrix += self.build_distance_matrix(response)
+        return distance_matrix
+
+    def prepare_destinations(self):
+        ''' Inherit this function in your model to provide the location
+            parameters for the Google Directions API request
+            https://developers.google.com/maps/documentation/directions/intro#RequestParameters
+            :return list: a list of strings representing the destination
+             address for each record in the recordset
+        '''
+        destinations = []
+        return destinations
+
+    def get_route(self, manager, routing, solution):
+        ''' Get route and durations between route nodes from a solution.
+           Return a dictionary of the lists generated.'''
+        index = routing.Start(0)
+        vehicle_id = 0
+        route = [manager.IndexToNode(index)]
+        trip_durations = []
+        while not routing.IsEnd(index):
+            previous_index = index
+            index = solution.Value(routing.NextVar(index))
+            # Build sorted route node list
+            route.append(manager.IndexToNode(index))
+            # Create list of durations between route nodes
+            trip_durations.append(
+                routing.GetArcCostForVehicle(
+                    previous_index, index, vehicle_id)
+            )
+        return {
+            'route': route,
+            'durations': trip_durations,
+        }
+
+    @api.multi
+    def optimize_records(self, origin):
+        ''' This is the primary function to initiate the geo optimization of a
+           recordset. You must provide an origin string parameter in the same
+           format as required in the prepare_destinations function.
+           Returns a dictionary with a list of the record ids sorted in
+           the optimized order and a list of durations between nodes
+        '''
+        if not origin:
+            raise ValidationError(_('No origin defined for route optimization'))
+        unsorted_ids = self.ids
+        num_vehicles = 1
+        # A depot (start & end point) is the first index in `[addresses]`
+        depot = 0
```

Comment:
Shouldn't it be a constant?

It seems it's value is always `0`

Reply:
I made it a constant, local to that function.  It shouldn't change

In the future, `DEPOT` may be replaced with a separate start node and end node as the library being used allows this